File: backend/app/notifications/router.py
"""
Notifications Module — Router
"""
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.notifications.service import NotificationService

router = APIRouter()


def _serialize(notifications) -> list[dict]:
    return [
        {
            "id": str(n.id),
            "type": n.type,
            "title": n.title,
            "message": n.message,
            "severity": n.severity,
            "is_read": n.is_read,
            "created_at": n.created_at.isoformat() if n.created_at else None,
        }
        for n in notifications
    ]


@router.get("")
async def list_notifications(
    limit: int = Query(50, ge=1, le=200),
    session: AsyncSession = Depends(get_db),
):
    """List in-app notifications (unread first, newest first)."""
    service = NotificationService(session)
    return {
        "notifications": _serialize(await service.list_notifications(limit)),
        "unread": await service.unread_count(),
    }


@router.get("/summary")
async def notification_summary(session: AsyncSession = Depends(get_db)):
    """Unread count for the bell badge."""
    return {"unread": await NotificationService(session).unread_count()}


@router.post("/refresh")
async def refresh_notifications(session: AsyncSession = Depends(get_db)):
    """Recompute notifications from current data + enabled settings.

    Idempotent — deduped by fingerprint, safe to call on every app load.
    """
    return await NotificationService(session).refresh()


@router.post("/read-all")
async def mark_all_read(session: AsyncSession = Depends(get_db)):
    return {"updated": await NotificationService(session).mark_all_read()}


@router.delete("")
async def clear_notifications(session: AsyncSession = Depends(get_db)):
    """Archive all notifications."""
    return {"deleted": await NotificationService(session).clear_all()}


@router.post("/{notification_id}/read")
async def mark_read(
    notification_id: uuid.UUID,
    session: AsyncSession = Depends(get_db),
):
    return {"updated": await NotificationService(session).mark_read(notification_id)}

File: backend-rs/src/notifications/router.rs
//! Notifications HTTP routes, mirroring `backend/app/notifications/router.py`.
//!
//! GET    /api/v1/notifications          -> { notifications, unread }
//! DELETE /api/v1/notifications          -> archive all
//! GET    /api/v1/notifications/summary  -> { unread }
//! POST   /api/v1/notifications/refresh  -> recompute (idempotent)
//! POST   /api/v1/notifications/read-all -> mark all read
//! POST   /api/v1/notifications/{id}/read

use axum::extract::{Path, Query, State};
use axum::routing::{get, post};
use axum::{Json, Router};
use serde::Deserialize;
use serde_json::{json, Value};
use uuid::Uuid;

use crate::auth::extractors::AuthUser;
use crate::core::error::AppError;
use crate::core::state::AppState;
use crate::notifications::models::Notification;
use crate::notifications::service;

#[derive(Debug, Default, Deserialize)]
struct ListQuery {
    limit: Option<i64>,
}

fn serialize(n: &Notification) -> Value {
    json!({
        "id": n.id,
        "type": n.r#type,
        "title": n.title,
        "message": n.message,
        "severity": n.severity,
        "is_read": n.is_read,
        "created_at": n.created_at,
    })
}

pub fn build_router() -> Router<AppState> {
    Router::new()
        .route("/", get(list_notifications).delete(clear_notifications))
        .route("/summary", get(summary))
        .route("/refresh", post(refresh))
        .route("/read-all", post(read_all))
        .route("/{id}/read", post(mark_read))
}

async fn list_notifications(
    State(state): State<AppState>,
    _auth: AuthUser,
    Query(q): Query<ListQuery>,
) -> Result<Json<Value>, AppError> {
    let limit = q.limit.unwrap_or(50).clamp(1, 200);
    let items = service::list(&state.pool, limit).await?;
    let notifications: Vec<Value> = items.iter().map(serialize).collect();
    Ok(Json(json!({
        "notifications": notifications,
        "unread": service::unread_count(&state.pool).await?,
    })))
}

async fn summary(State(state): State<AppState>, _auth: AuthUser) -> Result<Json<Value>, AppError> {
    Ok(Json(json!({ "unread": service::unread_count(&state.pool).await? })))
}

async fn refresh(State(state): State<AppState>, _auth: AuthUser) -> Result<Json<Value>, AppError> {
    let result = service::refresh(&state.pool).await?;
    Ok(Json(result))
}

async fn read_all(State(state): State<AppState>, _auth: AuthUser) -> Result<Json<Value>, AppError> {
    Ok(Json(json!({ "updated": service::mark_all_read(&state.pool).await? })))
}

async fn clear_notifications(
    State(state): State<AppState>,
    _auth: AuthUser,
) -> Result<Json<Value>, AppError> {
    Ok(Json(json!({ "deleted": service::clear_all(&state.pool).await? })))
}

async fn mark_read(
    State(state): State<AppState>,
    _auth: AuthUser,
    Path(id): Path<Uuid>,
) -> Result<Json<Value>, AppError> {
    Ok(Json(json!({ "updated": service::mark_read(&state.pool, &id.to_string()).await? })))
}
